File: app/src/main/java/com/verbatoria/business/schedule/datasource/ScheduleDataSource.java
package com.verbatoria.business.schedule.datasource;

import com.verbatoria.business.schedule.models.ScheduleItemModel;
import com.verbatoria.utils.DateUtils;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.WeakHashMap;

import static com.verbatoria.utils.LocaleHelper.LOCALE_RU;

/**
 * @author nikitaremnev
 */

public class ScheduleDataSource implements IScheduleDataSource<String, Date, Date, ScheduleItemModel> {

    private static final int ROWS_COUNT = 8;
    private static final int START_HOUR = 8;
    private static final int END_HOUR = 19;
    private static final int WEEK_COUNT = 7;
    private static final int FIRST_DAY_OF_WEEK = 1;
    private static final int COLUMN_COUNT = END_HOUR - START_HOUR + 1;

    private final Map<Integer, List<ScheduleItemModel>> mItems = new WeakHashMap<>();

    private Calendar mCalendar = Calendar.getInstance(new Locale(LOCALE_RU));

    private Calendar mOriginalCalendar;

    public ScheduleDataSource() {
        mOriginalCalendar = Calendar.getInstance(new Locale(LOCALE_RU));
    }

    public ScheduleDataSource(Calendar calendar) {
        mOriginalCalendar = calendar;
        int dayOfWeekIndex = 1;
        while (dayOfWeekIndex < ROWS_COUNT) {
            List<ScheduleItemModel> dayGenerated = new ArrayList<>();
            for (int i = START_HOUR; i < END_HOUR; i++) {
                dayGenerated.add(new ScheduleItemModel());
            }
            mItems.put(dayOfWeekIndex, dayGenerated);
            dayOfWeekIndex ++;
        }
    }

    @Override
    public int getRowsCount() {
        return ROWS_COUNT;
    }

    @Override
    public int getColumnsCount() {
        return COLUMN_COUNT;
    }

    @Override
    public String getFirstHeaderData() {
        return null;
    }

    @Override
    public Date getRowHeaderData(int index) {
        mCalendar = Calendar.getInstance(new Locale(LOCALE_RU));
        mCalendar.setTime(mOriginalCalendar.getTime());
        int firstDayOfWeek = getFirstDayOfWeek();
        mCalendar.set(Calendar.DAY_OF_YEAR, mCalendar.get(Calendar.DAY_OF_YEAR) - (getCurrentDayOfWeek() - firstDayOfWeek - (index - 1)));
        return mCalendar.getTime();
    }

    @Override
    public Date getColumnHeaderData(int index) {
        mCalendar.set(Calendar.HOUR_OF_DAY, START_HOUR + index - 1);
        mCalendar.set(Calendar.MINUTE, 0);
        return mCalendar.getTime();
    }

    @Override
    public ScheduleItemModel getItemData(int rowIndex, int columnIndex) {
        return mItems.get(rowIndex).get(columnIndex - 1);
    }

    @Override
    public Date getNextWeekStart() {
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT);
        Date result = getWeekStart();
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT);
        return result;
    }

    @Override
    public Date getNextWeekFinish() {
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT);
        Date result = getWeekEnd();
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT);
        return result;
    }

    @Override
    public Date getPreviousWeekStart() {
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT);
        Date result = getWeekStart();
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT);
        return result;
    }

    @Override
    public Date getPreviousWeekFinish() {
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT);
        Date result = getWeekEnd();
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT);
        return result;
    }

    @Override
    public String getWeekTitle() {
        return DateUtils.datePeriodToString(getRowHeaderData(FIRST_DAY_OF_WEEK), getRowHeaderData(ROWS_COUNT - 1));
    }

    @Override
    public void clearSchedule() {
        int dayOfWeekIndex = FIRST_DAY_OF_WEEK;
        while (dayOfWeekIndex < ROWS_COUNT) {
            List<ScheduleItemModel> scheduleItemModelList = mItems.get(dayOfWeekIndex);
            for (int i = 0; i < scheduleItemModelList.size(); i++) {
                scheduleItemModelList.get(i).setSelected(false);
            }
            dayOfWeekIndex ++;
        }
    }

    @Override
    public Date getWeekStart() {
        mCalendar = Calendar.getInstance(new Locale(LOCALE_RU));
        mCalendar.setTime(mOriginalCalendar.getTime());
        int firstDayOfWeek = getFirstDayOfWeek();
        mCalendar.set(Calendar.HOUR_OF_DAY, 0);
        mCalendar.set(Calendar.MINUTE, 0);
        mCalendar.set(Calendar.DAY_OF_YEAR, mCalendar.get(Calendar.DAY_OF_YEAR) - (getCurrentDayOfWeek() - firstDayOfWeek));
        return mCalendar.getTime();
    }

    @Override
    public Date getWeekEnd() {
        mCalendar = Calendar.getInstance(new Locale(LOCALE_RU));
        mCalendar.setTime(mOriginalCalendar.getTime());
        int firstDayOfWeek = getFirstDayOfWeek();
        mCalendar.set(Calendar.HOUR_OF_DAY, 0);
        mCalendar.set(Calendar.MINUTE, 0);
        mCalendar.set(Calendar.DAY_OF_YEAR, mCalendar.get(Calendar.DAY_OF_YEAR) - (getCurrentDayOfWeek() - firstDayOfWeek - WEEK_COUNT));
        return mCalendar.getTime();
    }

    @Override
    public Date getDeleteBorder(int weeksForwardCount) {
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT * weeksForwardCount);
        Date result = getWeekEnd();
        mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT * weeksForwardCount);
        return result;
    }

    @Override
    public Calendar getOriginalCalendar() {
        return mOriginalCalendar;
    }

    @Override
    public void setWorkingInterval(Date date) {
        mCalendar = Calendar.getInstance(new Locale(LOCALE_RU));
        mCalendar.setTime(date);
        int currentHour = mCalendar.get(Calendar.HOUR_OF_DAY);
        if (isHourInBorders(currentHour)) {
            mItems.get(getCurrentDayOfWeek()).get(currentHour - START_HOUR).setSelected(true);
        }
    }

    @Override
    public Map<Date, List<Date>> getItems(boolean selected) {
        int dayOfWeekIndex = FIRST_DAY_OF_WEEK;
        Map<Date, List<Date>> resultItems = new HashMap<>();
        while (dayOfWeekIndex < ROWS_COUNT) {
            List<ScheduleItemModel> scheduleItemModelList = mItems.get(dayOfWeekIndex);
            List<Date> subItems = new ArrayList<>();
            for (int i = 0; i < scheduleItemModelList.size(); i++) {
                if (scheduleItemModelList.get(i).isSelected() == selected) {
                    subItems.add(getColumnHeaderData(i + 1));
                }
            }
            resultItems.put(getRowHeaderData(dayOfWeekIndex), subItems);
            dayOfWeekIndex ++;
        }
        return resultItems;
    }

    @Override
    public Map<Date, List<Date>> getItems(boolean selected, int nextWeeksCount) {
        Map<Date, List<Date>> resultItems = new HashMap<>();
        for (int weekForward = 0; weekForward < nextWeeksCount + 1; weekForward ++) {
            mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT * weekForward);
            int dayOfWeekIndex = FIRST_DAY_OF_WEEK;
            while (dayOfWeekIndex < ROWS_COUNT) {
                List<ScheduleItemModel> scheduleItemModelList = mItems.get(dayOfWeekIndex);
                List<Date> subItems = new ArrayList<>();
                for (int i = 0; i < scheduleItemModelList.size(); i++) {
                    if (scheduleItemModelList.get(i).isSelected() == selected) {
                        subItems.add(getColumnHeaderData(i + 1));
                    }
                }
                resultItems.put(getRowHeaderData(dayOfWeekIndex), subItems);
                dayOfWeekIndex ++;
            }
            mOriginalCalendar.set(Calendar.DAY_OF_YEAR, mOriginalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT * weekForward);
        }
        return resultItems;
    }

    private int getFirstDayOfWeek() {
        return mCalendar.getFirstDayOfWeek() - 1;
    }

    private int getCurrentDayOfWeek() {
        return mCalendar.get(Calendar.DAY_OF_WEEK) == FIRST_DAY_OF_WEEK ? WEEK_COUNT : mCalendar.get(Calendar.DAY_OF_WEEK) - 1;
    }

    private boolean isHourInBorders(int hour) {
        return hour >= START_HOUR && hour < END_HOUR;
    }

}

File: app/src/main/java/com/verbatoria/domain/schedule/model/ScheduleDataSource.kt
package com.verbatoria.domain.schedule.model

import com.verbatoria.business.dashboard.LocalesAvailable
import com.verbatoria.domain.schedule.model.ScheduleDataSource.Companion.COLUMN_COUNT
import com.verbatoria.domain.schedule.model.ScheduleDataSource.Companion.ROWS_COUNT
import com.verbatoria.infrastructure.extensions.dropToStartOfTheDay
import com.verbatoria.infrastructure.extensions.formatToShortDate
import java.util.*

/**
 * @author n.remnev
 */

private const val START_HOUR = 8
private const val END_HOUR = 19
private const val WEEK_COUNT = 7
private const val FIRST_DAY_OF_WEEK = 1

interface ScheduleDataSource {

    companion object {

        const val ROWS_COUNT = 8

        const val COLUMN_COUNT = 12

    }

    fun getRowsCount(): Int

    fun getColumnsCount(): Int

    fun getScheduleCellItem(row: Int, column: Int): ScheduleCellItem?

    fun getFirstDayOfNextWeek(): Date

    fun getLastDayOfNextWeek(): Date

    fun getFirstDayOfCurrentWeek(): Date

    fun getLastDayOfCurrentWeek(): Date

    fun getFirstDayOfPreviousWeek(): Date

    fun getLastDayOfPreviousWeek(): Date

    fun getCurrentWeekTitle(): String

    fun getRowHeaderDate(row: Int): Date

    fun getColumnHeaderDate(column: Int): Date

    fun getSchedule(): Map<Int, List<ScheduleCellItem>>

    fun setWorkingInterval(date: Date)

    fun clear()

}

class ScheduleDataSourceImpl : ScheduleDataSource {

    private var originalCalendar: Calendar
    private var currentCalendar: Calendar = Calendar.getInstance(Locale(LocalesAvailable.RUSSIAN_LOCALE))

    private var scheduleItems: MutableMap<Int, List<ScheduleCellItem>> = hashMapOf()

    constructor() {
        originalCalendar = Calendar.getInstance(Locale(LocalesAvailable.RUSSIAN_LOCALE))
        originalCalendar.dropToStartOfTheDay()

        var dayOfWeekIndex = 1
        while (dayOfWeekIndex < ROWS_COUNT) {
            val dayGenerated = mutableListOf<ScheduleCellItem>()
            for (i in START_HOUR until END_HOUR) {
                dayGenerated.add(
                    ScheduleCellItem(
                        startHour = i,
                        isSelected = false
                    )
                )
            }
            scheduleItems[dayOfWeekIndex] = dayGenerated
            dayOfWeekIndex++
        }
    }

    constructor(calendar: Calendar) {
        originalCalendar = calendar
        originalCalendar.dropToStartOfTheDay()

        var dayOfWeekIndex = 1
        while (dayOfWeekIndex < ROWS_COUNT) {
            val dayGenerated = mutableListOf<ScheduleCellItem>()
            for (i in START_HOUR until END_HOUR) {
                dayGenerated.add(
                    ScheduleCellItem(
                        startHour = i,
                        isSelected = false
                    )
                )
            }
            scheduleItems[dayOfWeekIndex] = dayGenerated
            dayOfWeekIndex++
        }
    }

    //region ScheduleDataSource

    override fun getRowsCount(): Int = ROWS_COUNT

    override fun getColumnsCount(): Int = COLUMN_COUNT

    override fun getScheduleCellItem(row: Int, column: Int): ScheduleCellItem? =
        scheduleItems[row]?.get(column - 1)

    override fun getFirstDayOfNextWeek(): Date {
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT
        )
        val result = getFirstDayOfCurrentWeek()
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT
        )
        return result

    }

    override fun getLastDayOfNextWeek(): Date {
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT
        )
        val result = getLastDayOfCurrentWeek()
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT
        )
        return result
    }

    override fun getFirstDayOfCurrentWeek(): Date {
        currentCalendar.time = originalCalendar.time
        val firstDayOfWeek = getFirstDayOfWeek()
        currentCalendar.dropToStartOfTheDay()
        currentCalendar.set(
            Calendar.DAY_OF_YEAR,
            currentCalendar.get(Calendar.DAY_OF_YEAR) - (getCurrentDayOfWeek() - firstDayOfWeek)
        )
        return currentCalendar.time
    }

    override fun getLastDayOfCurrentWeek(): Date {
        currentCalendar.time = originalCalendar.time
        val firstDayOfWeek = getFirstDayOfWeek()
        currentCalendar.dropToStartOfTheDay()
        currentCalendar.set(
            Calendar.DAY_OF_YEAR,
            currentCalendar.get(Calendar.DAY_OF_YEAR) - (getCurrentDayOfWeek() - firstDayOfWeek - WEEK_COUNT)
        )
        return currentCalendar.time
    }

    override fun getFirstDayOfPreviousWeek(): Date {
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT
        )
        val result = getFirstDayOfCurrentWeek()
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT
        )
        return result
    }

    override fun getLastDayOfPreviousWeek(): Date {
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) - WEEK_COUNT
        )
        val result = getLastDayOfCurrentWeek()
        originalCalendar.set(
            Calendar.DAY_OF_YEAR,
            originalCalendar.get(Calendar.DAY_OF_YEAR) + WEEK_COUNT
        )
        return result
    }

    override fun getCurrentWeekTitle(): String =
        getRowHeaderDate(FIRST_DAY_OF_WEEK).formatToShortDate() + " - " +
                getRowHeaderDate(ROWS_COUNT - 1).formatToShortDate()

    override fun getRowHeaderDate(row: Int): Date {
        currentCalendar.time = originalCalendar.time
        val firstDayOfWeek = getFirstDayOfWeek()
        currentCalendar.set(
            Calendar.DAY_OF_YEAR,
            currentCalendar.get(Calendar.DAY_OF_YEAR) - (getCurrentDayOfWeek() - firstDayOfWeek - (row - 1))
        )
        return currentCalendar.time
    }

    override fun getColumnHeaderDate(column: Int): Date {
        currentCalendar.set(Calendar.HOUR_OF_DAY, START_HOUR + column - 1)
        currentCalendar.set(Calendar.MINUTE, 0)
        return currentCalendar.time
    }

    override fun getSchedule(): Map<Int, List<ScheduleCellItem>> =
        scheduleItems

    override fun setWorkingInterval(date: Date) {
        currentCalendar = Calendar.getInstance(Locale(LocalesAvailable.RUSSIAN_LOCALE))
        currentCalendar.time = date
        val currentHour = currentCalendar.get(Calendar.HOUR_OF_DAY)
        if (isHourInBorders(currentHour)) {
            scheduleItems[getCurrentDayOfWeek()]?.let { scheduleCellItemsList ->
                scheduleCellItemsList[currentHour - START_HOUR].isSelected = true
            }
        }
    }

    override fun clear() {
        var dayOfWeekIndex = FIRST_DAY_OF_WEEK
        while (dayOfWeekIndex < ROWS_COUNT) {
            scheduleItems[dayOfWeekIndex]?.let { scheduleItem ->
                for (i in scheduleItem.indices) {
                    scheduleItem[i].isSelected = false
                }
            }
            dayOfWeekIndex++
        }
    }

    //endregion

    private fun getFirstDayOfWeek(): Int = currentCalendar.firstDayOfWeek - 1

    private fun getCurrentDayOfWeek(): Int =
        if (currentCalendar.get(Calendar.DAY_OF_WEEK) == FIRST_DAY_OF_WEEK)
            WEEK_COUNT
        else
            currentCalendar.get(Calendar.DAY_OF_WEEK) - 1

    private fun isHourInBorders(hour: Int): Boolean = hour in START_HOUR until END_HOUR

}